miners/text/template_miner.py

```diff
@@ -0,0 +1,643 @@
+    def backward ( self, pubkey:str, inputs_x:torch.FloatTensor, grads_dy:torch.FloatTensor, modality:int ) -> torch.FloatTensor:
+        r""" Subscribed to an axon servicing endpoint: Processes backward messages from the wire.
+            Arguments reflect an RPC backward request from another miner in the network, the response tensor
+            should be the gradients of the miner's nucleus w.r.t to the inputs_x and the passed output grads_dy.
+
+            Args:
+                pubkey ( str, `required`):
+                    The public key of the caller.
+                inputs_x ( :obj:`torch.Tensor`, `required`):
+                    torch inputs from previous forward call.
+                grads_dy ( :obj:`torch.Tensor`, `required`):
+                    torch grads of forward output.
+                modality ( bittensor.proto.Modality, `required`):
+                    modality of inputs e.g. bittensor.proto.Modality.TEXT.
+
+            Returns:
+                outputs (:obj:`torch.FloatTensor`, `optional`):
+                    The gradients w.r.t to the inputs [batch_size, sequence_len, -1]
+        """
+        if self.config.miner.compute_remote_gradients:
+            with torch.enable_grad():
+
+                # ---- Set up inputs for gradient computations.
+                inputs_x.requires_grad = True
+                inputs_x = inputs_x.to( self.device )
+                grads_dy = grads_dy.to( self.device )
+                outputs_y = self.nucleus.local_forward( inputs = inputs_x ).to( self.device )
+
+                # ---- The backward call will accumulate gradients on our parameters.
+                if self.config.miner.accumulate_remote_gradients:
+                    torch.autograd.backward (
+                        tensors = [outputs_y],
+                        grad_tensors = [grads_dy]
+                    )
+                    return inputs_x.grad if inputs_x.grad != None else None
+
+                # ---- The backward call will simply compute the gradients without accumulating them.
+                else:
+                    grads_dy = torch.autograd.grad (
+                        outputs = outputs_y,
+                        inputs = inputs_x,
+                        grad_outputs = grads_dy,
+                        only_inputs = True,
+                        create_graph = False,
+                        retain_graph = False
+                    )[0]
+                    return grads_dy
+
+        # if ! compute_remote_gradients, NO-OP.
+        else:
+            return None
+
+    def checkpoint( self ):
+        r""" Optionally Saves, updates and then reloads the miner training state.
+        """
+        last_saved = self.get_saved_state()
+        if last_saved == None or last_saved['epoch_loss'] >= self.epoch_loss:
+            self.save()
+        bittensor.neuron.metagraph.load()
+        bittensor.neuron.metagraph.sync()
+        bittensor.neuron.metagraph.save()
+        #self.reload()
```

Comment:
We should add a NAN check here, something like "have we nan'd out? yes = reload()"

Reply:
Fixed